### PR TITLE
[Small Feature] PSP: Spherical Neighborhood Queries

### PR DESCRIPTION
**Pull request by sgiraudot**
## Rationale

Many _Point Set Processing_ algorithms use a KD-tree to make neighborhood queries. These algorithms currently use a fixed K number of nearest neighbors, but in some cases it can be interesting to use a fixed spherical neighborhood radius instead.

This small feature adds the possibility for 6 algorithms to use the (existing) named parameter `neighbor_radius` to switch to this behavior. The default behavior remains the same (using K).

## Summary of API changes

The following functions are impacted:
- `bilateral_smooth_point_set()`
- `jet_estimate_normals()`
- `jet_smooth_point_set()`
- `mst_orient_normals()`
- `pca_estimate_normals()`
- `remove_outliers()`

The named parameter `neighbor_radius` can be used. If this parameter is provided, then the parameter `k` is used as a maximum number of points (to avoid overly large number of neighbors when doing a sphere query on a very high density area). If `k==0`, no limit is used.

This is a non-breaking change, as the named parameter is optional and the default behavior remains the same (with the existing `k` parameter remaining unchanged).

The user manual is slightly modified to indicate when the user has the choice between KNN and radius. The example `normals_example.h` is modified to show how to switch to the radius behavior.

## License and copyright ownership

(no change)

## CHANGES.md

TODO

## Status

-   Developed and locally tested (GNU/Linux)
-   (Empty) small feature page: [PSP Spherical Neighborhood Queries](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/PSP_Spherical_Neighborhood_Queries)

## Submission

* [__User manual:__](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/index.html#Chapter_Point_Set_Processing) only slightly modified, modifications are indicated in red boxes
* __Reference Manual:__ (the documentation of `neighbor_radius` is the same for all functions)
  - [`bilateral_smooth_point_set()`](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga4f82723e2f0bb33f3677e29e0208a256)
  - [`jet_estimate_normals()`](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga0cd0f87de690d4edf82740e856efa491)
  - [`jet_smooth_point_set()`](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga549402c0a8a8b6b71875181e93961521)
  - [`mst_orient_normals()`](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga50c98d5c5ae5535bce6f32eddbd03f33)
  - [`pca_estimate_normals()`](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga8c642da96a025ab32445aeb6cc219b0b)
  - [`remove_outliers()`](https://cgal.geometryfactory.com/~sgiraudot/PSP-Radius_v1/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#gafd0b5a21ec5042e4bca09cb43f1847f9)

### CODE REVIEW

**Comment by afabri** [comment]
I have read the updated documentation and pre-approve.

**Comment by maxGimeno** [comment]
[Catastrophe](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-17.shtml#Point_set_processing_3) :(

File: Point_set_processing_3/include/CGAL/Point_set_processing_3/internal/neighbor_query.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  typedef typename Neighbor_search::iterator Search_iterator;
+  typedef CGAL::Fuzzy_sphere<TreeTraits> Sphere;
+  
+  if (neighbor_radius != FT(0))
+  {
+    Sphere fs (query, neighbor_radius, 0, tree.traits());
+
+    // if k=0, no limit on the number of neighbors returned
+    if (k == 0)
+      k = std::numeric_limits<unsigned int>::max();
```
Comment (afabri) [inline Point_set_processing_3/include/CGAL/Point_set_processing_3/internal/neighbor_query.h]: You have to replace it with `(std::numeric_limits<unsigned int>::max)()`

**Comment by afabri** [comment]
@maxGimeno please add it to the testsuite.   Also it is pre-approved + 14 days => approved.

**Comment by maxGimeno** [comment]
It has been in the testsuite for 17 days.

**Comment by maxGimeno** [comment]
Successfully testsed in [Ic-29](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-29.shtml)

**Comment by lrineau** [comment]
This PR has been modified after being tested. I remove the "Tested" label.

**Comment by afabri** [comment]
I reviewed the text in the [latest change](https://github.com/CGAL/cgal/pull/3869/commits/6b112dda54ffb9839833b0e6c31bd32dfe7493ec), and it is correct.

**Comment by lrineau** [comment]
I have merged this PR, but the change to `CHANGES.md` is still TODO.

**Comment by lrineau** [comment]
@sgiraudot This PR still does not have any corresponding announcement in the `CHANGES.md`.

**Pull request merged**